### PR TITLE
Improve uniform locations part 1

### PR DESCRIPTION
**Pull request by FutureMaximus**
Replaces uniform locations solely using integers with descriptive uniform locations using the ShaderLocations class for View3DDeferredRenderer and GameClientWindow.

### CODE REVIEW

File: FGEGraphics/ClientSystem/ShaderLocations.cs [hunk 1]
```diff
@@ -6,6 +6,7 @@
 // hold any right or permission to use this software until such time as the official license is identified.
 //
 
+using FGEGraphics.ClientSystem.ViewRenderSystem;
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ShaderLocations.cs]: accidental import?

File: FGEGraphics/ClientSystem/ShaderLocations.cs [hunk 2]
```diff
@@ -42,4 +46,309 @@ public static class Common2D
...
+        /// The HDR exposure value.
+        /// Uniform: hdrExposure
+        /// </summary>
+        public const int HDR_EXPOSURE = 10;
+
+        /// <summary>
+        /// Camera position relative to rendering origin.
+        /// Uniform: cameraPos
+        /// </summary>
+        public const int CAMERA_RELATIVE_POSITION = 14;
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ShaderLocations.cs]: `relative` is redundant here, _everything_ in the graphics end is relative to the rendering origin, naturally

EDIT: oh i see that was copied from pre-existing docs lol.
Still for this field name drop that bit

File: FGEGraphics/ClientSystem/ShaderLocations.cs [hunk 3]
```diff
@@ -42,4 +46,309 @@ public static class Common2D
...
+        public const int Z_DISTANCE = 17;
+
+        /// <summary>
+        /// What color any fog to apply is. For no fog, the alpha value will be zero.
+        /// Uniform: fogCol
+        /// </summary>
+        public const int FOG_COLOR = 18;
+
+        /// <summary>
+        /// How much to desaturation the view by. 1.0 = fully desaturated.
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ShaderLocations.cs]: typo

File: FGEGraphics/ClientSystem/ShaderLocations.cs (hunk 1 above)
Comment (FutureMaximus) [reply FGEGraphics/ClientSystem/ShaderLocations.cs]: Yes my bad I must have used a see class thing in the summary at one point but removed it

File: FGEGraphics/ClientSystem/ShaderLocations.cs (hunk 2 above)
Comment (FutureMaximus) [reply FGEGraphics/ClientSystem/ShaderLocations.cs]: I just followed what the comment said in the shader.

File: FGEGraphics/ClientSystem/ShaderLocations.cs (hunk 3 above)
Comment (FutureMaximus) [reply FGEGraphics/ClientSystem/ShaderLocations.cs]: I copied that one from the shader.

File: FGEGraphics/ClientSystem/ShaderLocations.cs [hunk 4]
```diff
@@ -42,4 +46,309 @@ public static class Common2D
...
+
+        /// <summary>
+        /// Distance that determines whether a pixel is in the sky or not.
+        /// Uniform: SKY_DIST
+        /// </summary>
+        public const int SKY_DISTANCE = 16;
+    }
+
+    /// <summary>Locations used in the finalgodray shader.</summary>
+    public static class FinalGodray
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ShaderLocations.cs]: `FinalPass`

File: FGEGraphics/ClientSystem/ShaderLocations.cs [hunk 5]
```diff
@@ -42,4 +46,309 @@ public static class Common2D
...
+        public const int MAX_DEPTH = 15;
+
+        /// <summary>
+        /// Distance that determines whether a pixel is in the sky or not.
+        /// Uniform: SKY_DIST
+        /// </summary>
+        public const int SKY_DISTANCE = 16;
+    }
+
+    /// <summary>Locations used in the finalgodray shader.</summary>
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ShaderLocations.cs]: Deferred-shader specifics should all be in a Deferred subsection, separate from a Forward section

File: FGEGraphics/ClientSystem/ShaderLocations.cs [hunk 6]
```diff
@@ -42,4 +46,309 @@ public static class Common2D
...
+        public const int FOG_COLOR = 18;
+
+        /// <summary>
+        /// How much to desaturation the view by. 1.0 = fully desaturated.
+        /// Uniform: desaturationAmount
+        /// </summary>
+        public const int DESATURATION = 19;
+
+        /// <summary>
+        /// What position the eye of the 3D camera view is at in the world.
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ShaderLocations.cs]: This is also copied from pre-existing docs but said pre-existing docs were ... not very clear. I think this is for 3DVR multi-eye rendering?

File: FGEGraphics/ClientSystem/ShaderLocations.cs [hunk 7]
```diff
@@ -42,4 +46,309 @@ public static class Common2D
...
+        public const int SCREEN_WIDTH = 24;
+
+        /// <summary>
+        /// How tall the screen is.
+        /// Uniform: height
+        /// </summary>
+        public const int SCREEN_HEIGHT = 25;
+
+        /// <summary>
+        /// A timer value, in seconds. Simply used for things that move.
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ShaderLocations.cs]: This is not a clear/useful doc, should reference _GlobalTickTime_ here

File: FGEGraphics/ClientSystem/ShaderLocations.cs [hunk 8]
```diff
@@ -42,4 +46,309 @@ public static class Common2D
...
+        public const int TIME = 26;
+
+        /// <summary>
+        /// How much motion blur to apply, and in what direction.
+        /// Uniform: mot_blur
+        /// </summary>
+        public const int MOTION_BLUR = 27;
+
+        /// <summary>
+        /// Whether to gray-scale the view.
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ShaderLocations.cs]: Should be clearer that it's a float-bool (0 for off 1 for on)

File: FGEGraphics/ClientSystem/ShaderLocations.cs [hunk 9]
```diff
@@ -42,4 +46,309 @@ public static class Common2D
...
+        /// Uniform: do_grayscale
+        /// </summary>
+        public const int DO_GRAYSCALE = 28;
+    }
+
+    /// <summary>Locations used in the transponly shader.</summary>
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ShaderLocations.cs]: _transparents-only_ render pass shader

File: FGEGraphics/ClientSystem/ShaderLocations.cs [hunk 10]
```diff
@@ -42,4 +46,309 @@ public static class Common2D
...
+        public const int DESATURATION = 4;
+
+        /// <summary>
+        /// Screen size.
+        /// Uniform: u_screen_size
+        /// </summary>
+        public const int SCREEN_SIZE = 8;
+
+        /// <summary>
+        /// Helper for lights used.
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ShaderLocations.cs]: not a helpful doc

File: FGEGraphics/ClientSystem/ShaderLocations.cs [hunk 11]
```diff
@@ -42,4 +46,309 @@ public static class Common2D
...
+        public const int FOG_DISTANCE = 13;
+
+        /// <summary>
+        /// Camera position, relative to rendering origin.
+        /// Uniform: cameraPos
+        /// </summary>
+        public const int CAMERA_RELATIVE_POSITION = 14;
+    }
+
+    /// <summary>Locations used in the transpadder shader.</summary>
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ShaderLocations.cs]: _transparents data adder_

File: FGEGraphics/ClientSystem/ShaderLocations.cs [hunk 12]
```diff
@@ -42,4 +46,309 @@ public static class Common2D
...
+    public static class TranspAdder
+    {
+        /// <summary>
+        /// The amount of lights used.
+        /// Uniform: lcount
+        /// </summary>
+        public const int LIGHTS_USED = 3;
+    }
+
+    /// <summary>Locations used in the clearer shader.</summary>
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ShaderLocations.cs]: This... can probably just not be added here. It's a very special case / experiment thing that maybe probably shouldn't exist anyway (same re `FPass`)

File: FGEGraphics/ClientSystem/ShaderLocations.cs [hunk 13]
```diff
@@ -42,4 +46,309 @@ public static class Common2D
...
+
+        /// <summary>
+        /// Whether to use a square root operation to adjust if the sub light is an orthographic light.
+        /// Uniform: should_sqrt
+        /// </summary>
+        public const int SHOULD_SQRT = 5;
+    }
+
+    /// <summary>Locations used in the FBO shader.</summary>
+    public static class FBO
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ShaderLocations.cs]: GBuffer

File: FGEGraphics/ClientSystem/ShaderLocations.cs [hunk 14]
```diff
@@ -42,4 +48,300 @@ public static class Common2D
         /// <summary>The rotation effect to apply.</summary>
         public const int ROTATION = 4;
     }
+
+    /// ===== FORWARD RENDERING SHADERS =====
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ShaderLocations.cs]: ... What.

Comment (FutureMaximus) [reply FGEGraphics/ClientSystem/ShaderLocations.cs]: Well it was this or putting them in a forward/deferred subclass meaning I'd have to rework all of the deferred locations (if there is a way I can just do that part automatically that'd be great to know)

Comment (FutureMaximus) [reply FGEGraphics/ClientSystem/ShaderLocations.cs]: But in the summaries, it's clear what they belong to.

Comment (mcmonkey4eva) [reply FGEGraphics/ClientSystem/ShaderLocations.cs]: Yes it should be in a subclass

Comment (mcmonkey4eva) [reply FGEGraphics/ClientSystem/ShaderLocations.cs]: re automation of it, just, uh, find-and-replace-in-files? VS Code has a really clean interface to do that quick, though VS can too

File: FGEGraphics/ClientSystem/ViewRenderSystem/View3DDeferredRenderer.cs [hunk 15]
```diff
@@ -102,22 +102,22 @@ public void RenderPass_Shadows()
                                 Patches.ShadowLightPatch?.Invoke(light, subLight);
                                 Shaders.Deferred.ShadowPass_Particles = Shaders.Deferred.ShadowPass_Particles.Bind();
                                 View.SetMatrix(2, Matrix4d.Identity);
-                                GL.Uniform1(5, (subLight is LightOrtho) ? 1.0f : 0.0f);
-                                GL.Uniform1(4, subLight.TransparentShadows ? 1.0f : 0.0f);
-                                GL.Uniform3(14, State.CameraRelativePosition);
+                                GL.Uniform1(ShaderLocations.Deferred.Shadow.SHOULD_SQRT, (subLight is LightOrtho) ? 1.0f : 0.0f); ;
```
Comment (mcmonkey4eva) [inline FGEGraphics/ClientSystem/ViewRenderSystem/View3DDeferredRenderer.cs]: typo on end of line

File: FGEGraphics/GraphicsHelpers/Models/ShapeGenerators.cs [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+using System;
+using System.Collections.Generic;
+using OpenTK.Mathematics;
+
+namespace FGEGraphics.GraphicsHelpers.Models;
+
+/// <summary>
+/// Generates shapes and returns them as a model.
+/// </summary>
+public static class ShapeGenerators
```
Comment (mcmonkey4eva) [inline FGEGraphics/GraphicsHelpers/Models/ShapeGenerators.cs]: This is an entirely unrelated addition and should not be in this PR
